From 34050b6a9030ec2bba777869d91c23f0ffcbac23 Mon Sep 17 00:00:00 2001
From: Henri Verbeet <hverbeet@locutus.nl>
Date: Sat, 6 Dec 2025 20:01:59 +0100
Subject: [PATCH 15/34] wined3d: Do not set "context_gl->needs_set" in
 wined3d_context_gl_release().

Commit ce44dd0ba41b88f6b8eb626863c3358b75aca66c got rid of
wined3d_context_gl_restore_pixel_format(), but seemingly inverted the
condition under which wined3d_context_gl_release() set the "needs_set"
flag. Previously the flag was set when
wined3d_context_gl_restore_pixel_format() changed the pixel format, and
now it's set when it doesn't. I.e., always.

Setting the "needs_set" flag causes wined3d_context_gl_activate() to
call wined3d_context_gl_set_gl_context(), which sets the pixel format and
makes the GL context current. These are expensive operations. This
commit improves performance in the rthdribl demo from a fairly
disappointing 1-2 fps to a more reasonable 70-80 fps on my Intel Skylake
setup. On my AMD gfx1200 setup the difference is less dramatic but still
quite significant, going from about 250 fps to about 350 fps.

It's perhaps worth pointing out that after this commit things are still
some way off from what they were on Wine 10.0: 150-160 fps on the Intel
setup, and about 450 fps on the AMD setup.

Signed-off-by: sewn <sewn@disroot.org>
---
 dlls/wined3d/context_gl.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/wined3d/context_gl.c b/dlls/wined3d/context_gl.c
index 66391ed3cb4..8aeb2284f44 100644
--- a/dlls/wined3d/context_gl.c
+++ b/dlls/wined3d/context_gl.c
@@ -1591,8 +1591,6 @@ void wined3d_context_gl_release(struct wined3d_context_gl *context_gl)
 
     if (!--context_gl->level)
     {
-        context_gl->needs_set = 1;
-
         if (context_gl->restore_ctx)
         {
             TRACE("Restoring GL context %p on device context %p.\n", context_gl->restore_ctx, context_gl->restore_dc);
-- 
2.52.0

